File: etapa3-construcao(geracao-de-cod)/stack2/Reserva.java
package com.hotel.reserva.domain.entities;

import com.hotel.reserva.domain.enums.StatusReserva;
import jakarta.persistence.*;
import lombok.*;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.temporal.ChronoUnit;
import java.util.UUID;

@Entity
@Table(name = "reservas")
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@EqualsAndHashCode(of = "id")
public class Reserva {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "quarto_id", nullable = false)
    private Quarto quarto;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "hospede_id", nullable = false)
    private Hospede hospede;

    @Column(nullable = false)
    private LocalDate dataCheckIn;

    @Column(nullable = false)
    private LocalDate dataCheckOut;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private StatusReserva status;

    @Column(nullable = false, precision = 10, scale = 2)
    private BigDecimal valorTotal;

    @Column(nullable = false, updatable = false)
    private LocalDateTime criadoEm;

    @Column(nullable = false)
    private LocalDateTime atualizadoEm;

    @PrePersist
    protected void onCreate() {
        criadoEm = LocalDateTime.now();
        atualizadoEm = LocalDateTime.now();
    }

    @PreUpdate
    protected void onUpdate() {
        atualizadoEm = LocalDateTime.now();
    }

    @Builder
    public Reserva(Quarto quarto, Hospede hospede, LocalDate dataCheckIn,
                   LocalDate dataCheckOut, BigDecimal valorTotal) {
        this.quarto = quarto;
        this.hospede = hospede;
        this.dataCheckIn = dataCheckIn;
        this.dataCheckOut = dataCheckOut;
        this.valorTotal = valorTotal;
        this.status = StatusReserva.PENDENTE;
    }

    public long calcularNumeroDiarias() {
        return ChronoUnit.DAYS.between(dataCheckIn, dataCheckOut);
    }

    public void confirmar() {
        if (this.status != StatusReserva.PENDENTE) {
            throw new IllegalStateException("Apenas reservas pendentes podem ser confirmadas");
        }
        this.status = StatusReserva.CONFIRMADA;
        this.atualizadoEm = LocalDateTime.now();
    }

    public void cancelar() {
        if (this.status == StatusReserva.FINALIZADA || this.status == StatusReserva.CANCELADA) {
            throw new IllegalStateException("Não é possível cancelar esta reserva");
        }
        this.status = StatusReserva.CANCELADA;
        this.atualizadoEm = LocalDateTime.now();
    }

    public void checkIn() {
        if (this.status != StatusReserva.CONFIRMADA) {
            throw new IllegalStateException("Apenas reservas confirmadas podem fazer check-in");
        }
        this.status = StatusReserva.EM_ANDAMENTO;
        this.atualizadoEm = LocalDateTime.now();
    }

    public void checkOut() {
        if (this.status != StatusReserva.EM_ANDAMENTO) {
            throw new IllegalStateException("Apenas reservas em andamento podem fazer check-out");
        }
        this.status = StatusReserva.FINALIZADA;
        this.atualizadoEm = LocalDateTime.now();
    }

    public boolean podeSerAlterada() {
        return this.status == StatusReserva.PENDENTE || this.status == StatusReserva.CONFIRMADA;
    }

    public void atualizarDados(Quarto quarto, Hospede hospede, LocalDate dataCheckIn,
                              LocalDate dataCheckOut, BigDecimal valorTotal) {
        if (!podeSerAlterada()) {
            throw new IllegalStateException("Esta reserva não pode ser alterada");
        }
        
        if (quarto != null) this.quarto = quarto;
        if (hospede != null) this.hospede = hospede;
        if (dataCheckIn != null) this.dataCheckIn = dataCheckIn;
        if (dataCheckOut != null) this.dataCheckOut = dataCheckOut;
        if (valorTotal != null) this.valorTotal = valorTotal;
        
        this.atualizadoEm = LocalDateTime.now();
    }

    public void validar() {
        if (quarto == null) {
            throw new IllegalArgumentException("Quarto é obrigatório");
        }
        
        if (hospede == null) {
            throw new IllegalArgumentException("Hóspede é obrigatório");
        }
        
        if (dataCheckOut.isBefore(dataCheckIn) || dataCheckOut.isEqual(dataCheckIn)) {
            throw new IllegalArgumentException("Data de check-out deve ser posterior à data de check-in");
        }
        
        if (valorTotal == null || valorTotal.compareTo(BigDecimal.ZERO) <= 0) {
            throw new IllegalArgumentException("Valor total deve ser maior que zero");
        }
    }
}

File: etapa3-construcao(geracao-de-cod)/stack2/src/domain/entities/Reserva.ts
import { ReservaData, StatusReserva, ValidationResult } from '../../types';
import { v4 as uuidv4 } from 'uuid';

export class Reserva {
  readonly id: string;
  private _quartoId: string;
  private _hospedeId: string;
  private _dataCheckIn: Date;
  private _dataCheckOut: Date;
  private _status: StatusReserva;
  private _valorTotal: number;
  readonly criadoEm: Date;
  private _atualizadoEm: Date;

  constructor(params: {
    quartoId: string;
    hospedeId: string;
    dataCheckIn: Date;
    dataCheckOut: Date;
    valorTotal: number;
    status?: StatusReserva;
    id?: string;
    criadoEm?: Date;
    atualizadoEm?: Date;
  }) {
    this.id = params.id || uuidv4();
    this._quartoId = params.quartoId;
    this._hospedeId = params.hospedeId;
    this._dataCheckIn = params.dataCheckIn;
    this._dataCheckOut = params.dataCheckOut;
    this._status = params.status || StatusReserva.PENDENTE;
    this._valorTotal = params.valorTotal;
    this.criadoEm = params.criadoEm || new Date();
    this._atualizadoEm = params.atualizadoEm || new Date();
  }

  get quartoId(): string {
    return this._quartoId;
  }

  get hospedeId(): string {
    return this._hospedeId;
  }

  get dataCheckIn(): Date {
    return this._dataCheckIn;
  }

  get dataCheckOut(): Date {
    return this._dataCheckOut;
  }

  get status(): StatusReserva {
    return this._status;
  }

  get valorTotal(): number {
    return this._valorTotal;
  }

  get atualizadoEm(): Date {
    return this._atualizadoEm;
  }

  calcularNumeroDiarias(): number {
    const diffTime = Math.abs(this._dataCheckOut.getTime() - this._dataCheckIn.getTime());
    const diffDays = Math.ceil(diffTime / (1000 * 60 * 60 * 24));
    return diffDays;
  }

  confirmar(): void {
    if (this._status !== StatusReserva.PENDENTE) {
      throw new Error('Apenas reservas pendentes podem ser confirmadas');
    }
    this._status = StatusReserva.CONFIRMADA;
    this._atualizadoEm = new Date();
  }

  cancelar(): void {
    if (this._status === StatusReserva.FINALIZADA || this._status === StatusReserva.CANCELADA) {
      throw new Error('Não é possível cancelar esta reserva');
    }
    this._status = StatusReserva.CANCELADA;
    this._atualizadoEm = new Date();
  }

  checkIn(): void {
    if (this._status !== StatusReserva.CONFIRMADA) {
      throw new Error('Apenas reservas confirmadas podem fazer check-in');
    }
    this._status = StatusReserva.EM_ANDAMENTO;
    this._atualizadoEm = new Date();
  }

  checkOut(): void {
    if (this._status !== StatusReserva.EM_ANDAMENTO) {
      throw new Error('Apenas reservas em andamento podem fazer check-out');
    }
    this._status = StatusReserva.FINALIZADA;
    this._atualizadoEm = new Date();
  }

  podeSerAlterada(): boolean {
    return this._status === StatusReserva.PENDENTE || this._status === StatusReserva.CONFIRMADA;
  }

  atualizarDados(params: {
    quartoId?: string;
    hospedeId?: string;
    dataCheckIn?: Date;
    dataCheckOut?: Date;
    valorTotal?: number;
  }): void {
    if (!this.podeSerAlterada()) {
      throw new Error('Esta reserva não pode ser alterada');
    }

    if (params.quartoId !== undefined) this._quartoId = params.quartoId;
    if (params.hospedeId !== undefined) this._hospedeId = params.hospedeId;
    if (params.dataCheckIn !== undefined) this._dataCheckIn = params.dataCheckIn;
    if (params.dataCheckOut !== undefined) this._dataCheckOut = params.dataCheckOut;
    if (params.valorTotal !== undefined) this._valorTotal = params.valorTotal;
    
    this._atualizadoEm = new Date();
  }

  validar(): ValidationResult {
    const errors: string[] = [];

    if (!this._quartoId) {
      errors.push('Quarto é obrigatório');
    }

    if (!this._hospedeId) {
      errors.push('Hóspede é obrigatório');
    }

    if (this._dataCheckOut <= this._dataCheckIn) {
      errors.push('Data de check-out deve ser posterior à data de check-in');
    }

    if (this._valorTotal <= 0) {
      errors.push('Valor total deve ser maior que zero');
    }

    return {
      isValid: errors.length === 0,
      errors
    };
  }

  toData(): ReservaData {
    return {
      id: this.id,
      quartoId: this._quartoId,
      hospedeId: this._hospedeId,
      dataCheckIn: this._dataCheckIn,
      dataCheckOut: this._dataCheckOut,
      status: this._status,
      valorTotal: this._valorTotal,
      criadoEm: this.criadoEm,
      atualizadoEm: this._atualizadoEm
    };
  }

  static fromData(data: ReservaData): Reserva {
    return new Reserva({
      quartoId: data.quartoId,
      hospedeId: data.hospedeId,
      dataCheckIn: data.dataCheckIn,
      dataCheckOut: data.dataCheckOut,
      valorTotal: data.valorTotal,
      status: data.status,
      id: data.id,
      criadoEm: data.criadoEm,
      atualizadoEm: data.atualizadoEm
    });
  }
}
